File: crack.go
package main

import (
	"bufio"
	"crypto/md5"
	"crypto/sha1"
	"fmt"
	"os"
	"strings"
	"sync"
	"unicode"
)

func generateHash(password string, algorithm string) string {
	switch algorithm {
	case "sha1":
		h := sha1.New()
		h.Write([]byte(password))
		return fmt.Sprintf("%x", h.Sum(nil))
	case "md5":
		h := md5.New()
		h.Write([]byte(password))
		return fmt.Sprintf("%x", h.Sum(nil))
	default:
		panic(fmt.Sprintf("Unsupported algorithm: %s", algorithm))
	}
}

func sanitize(input string) string {
	var output []rune
	for _, r := range input {
		if unicode.IsPrint(r) && !unicode.IsSpace(r) {
			output = append(output, r)
		}
		if len(output) >= 30 {
			break
		}
	}
	return string(output)
}

func main() {
	var storedPasswordHash, passwordFile string
	fmt.Print("Enter the stored password hash: ")
	fmt.Scan(&storedPasswordHash)

	storedPasswordHash = strings.ToLower(storedPasswordHash)

	var algorithm string
	switch len(storedPasswordHash) {
	case 32:
		algorithm = "md5"
	case 40:
		algorithm = "sha1"
	default:
		panic("Unable to detect hash algorithm based on hash length")
	}

	fmt.Print("Enter the file name containing the list of passwords: ")
	fmt.Scan(&passwordFile)

	file, err := os.Open(passwordFile)
	if err != nil {
		panic(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	var wg sync.WaitGroup
	found := make(chan bool, 1)

	go func() {
		for scanner.Scan() {
			password := scanner.Text()
			password = strings.TrimSpace(password)
			displayPassword := sanitize(password)
			fmt.Printf("\r\033[KTesting password: %s", displayPassword)
			wg.Add(1)
			go func(password string) {
				defer wg.Done()
				passwordHash := generateHash(password, algorithm)
				if passwordHash == storedPasswordHash {
					fmt.Printf("\r\033[KPassword %s is correct\n", displayPassword)
					found <- true
				}
			}(password)
		}
		wg.Wait()
		close(found)
	}()

	if _, ok := <-found; !ok {
		fmt.Printf("\r\033[KNone of the passwords in the list are correct\n")
	}
}

File: crack.py
import hashlib
import concurrent.futures

# Generate hash for a password
def generate_hash(password, algorithm):
    if algorithm == 'sha1':
        h = hashlib.sha1()
    elif algorithm == 'md5':
        h = hashlib.md5()
    else:
        raise ValueError(f"Unsupported algorithm: {algorithm}")
    h.update(password.encode('utf-8'))
    return h.hexdigest()

# Get user input for stored password hash and read user passwords from a file
stored_password_hash = input("Enter the stored password hash: ")
if len(stored_password_hash) == 32:
    algorithm = 'md5'
elif len(stored_password_hash) == 40:
    algorithm = 'sha1'
else:
    raise ValueError("Unable to detect hash algorithm")

password_file = input("Enter the file name containing the list of passwords: ")

# Read the passwords from the file
with open(password_file, 'r', encoding='iso-8859-1') as f:
    passwords = [line.strip() for line in f.readlines()]

# Check each password and output the result
with concurrent.futures.ThreadPoolExecutor() as executor:
    results = []
    for password in passwords:
        results.append(executor.submit(generate_hash, password, algorithm))

    for i, result in enumerate(concurrent.futures.as_completed(results)):
        password_hash = result.result()
        if password_hash == stored_password_hash:
            print(f"Password {passwords[i]} is correct")
            break
    else:
        print("None of the passwords in the list are correct")
